core/templates/dev/head/pages/exploration_editor/editor_tab/StateSolution.js

```diff
@@ -0,0 +1,277 @@
+
+            $scope.tmpSolution = {};
+            $scope.tmpSolution.answerIsExclusive = false;
+            $scope.tmpSolution.correctAnswer = '';
+            $scope.tmpSolution.explanation = '';
+
+            $scope.addSolutionForm = {};
+
+            if ($scope.objectType === 'CodeString') {
+              stateSolutionService.displayed = {
+                correctAnswer: {
+                  code: '',
+                  output: '',
+                  evaluation: '',
+                  error: ''
+                }
+              };
+            }
+
+            $scope.saveSolution = function() {
+              // Close the modal and save it afterwards.
+              var answer = ($scope.correctAnswer !== null) ? (
+                $scope.correctAnswer) : $scope.tmpSolution.correctAnswer;
+              if ($scope.objectType === 'CodeString') {
+                answer = {
+                  code: $scope.tmpSolution.correctAnswer,
+                  output: '',
+                  evaluation: '',
+                  error: ''
+                };
+              } else if ($scope.objectType === 'UnicodeString') {
+                answer = {
+                  ascii: '',
+                  latex: $scope.tmpSolution.correctAnswer
+                };
+              }
+
+              $modalInstance.close({
+                solution: angular.copy(
+                  SolutionObjectFactory.createNew(
+                    $scope.tmpSolution.answerIsExclusive,
+                    answer,
+                    $scope.tmpSolution.explanation))
+              });
+            };
+
+            $scope.cancel = function() {
+              $modalInstance.dismiss('cancel');
+              alertsService.clearWarnings();
+            };
+          }
+        ]
+      }).result.then(function(result) {
+        var explorationId =
+          explorationContextService.getExplorationId();
+        var currentStateName =
+          editorContextService.getActiveStateName();
+        var state = explorationStatesService.getState(currentStateName);
+        var interactionId = stateInteractionIdService.savedMemento;
+        var rulesServiceName = (
+          angularNameService.getNameOfInteractionRulesService(interactionId));
+        // Inject RulesService dynamically.
+        var rulesService = $injector.get(rulesServiceName);
+        var correctAnswer = result.solution.correctAnswer;
+        try {
+          AnswerClassificationService.getMatchingClassificationResult(
+            explorationId, state, correctAnswer, true, rulesService)
+            .then(function (result) {
+              if (editorContextService.getActiveStateName() !== (
+              result.outcome.dest)) {
+                stateSolutionService.saveDisplayedValue();
+                $scope.solutionIsSpecified = (
+                  !!JSON.stringify(stateSolutionService.savedMemento));
+              } else {
+                alertsService.addInfoMessage('That solution does not lead ' +
```

Comment:
I'm not sure about this. I can understand an invalid solution causing an error here, but in the case where the rules don't match the solution, would it not be reasonable to allow the creator to save the solution (but this displays a warning that makes the exploration non-publishable) and then go back and fix the rules?